File: python/multicast/cliente_multicast.py
import socket
import struct
import sys

# OFFSET pessoal (secao 3.3 do roteiro) - deve ser o MESMO do servidor
OFFSET = 81

GRUPO = "230.0.0.1"
PORTA = 4446 + OFFSET

# Nome opcional só para identificar o cliente nos logs:
#   python3 cliente_multicast.py Aluno-1
nome = sys.argv[1] if len(sys.argv) > 1 else "Aluno"

with socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP) as cliente:
    cliente.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

    # ADAPTACAO PARA macOS
    # No macOS (e nos BSD em geral) o SO_REUSEADDR sozinho nao permite que dois
    # processos escutem a mesma porta: o segundo cliente falha com
    # "OSError: [Errno 48] Address already in use". O SO_REUSEPORT resolve.
    # O hasattr existe porque essa opcao nao existe no Windows.
    if hasattr(socket, "SO_REUSEPORT"):
        cliente.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)

    cliente.bind(("", PORTA))

    # Entra no grupo multicast. O formato "=4sl" produz exatamente os 8 bytes
    # da struct ip_mreq (endereço do grupo + interface). Sem o "=", o Python
    # usa alinhamento nativo e gera 16 bytes, o que pode ser recusado no macOS.
    mreq = struct.pack("=4sl", socket.inet_aton(GRUPO), socket.INADDR_ANY)
    cliente.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)

    print(f"[MULTICAST] {nome} entrou no grupo {GRUPO}:{PORTA}")
    print("[MULTICAST] Aguardando avisos... (Ctrl+C para sair)")
    print()

    try:
        while True:
            dados, endereco = cliente.recvfrom(1024)
            print(f"[{nome}] recebeu -> {dados.decode('utf-8')}")
    except KeyboardInterrupt:
        print()
        print(f"[MULTICAST] {nome} saiu do grupo.")

File: java/multicast/ClienteMulticast.java
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.util.Collections;

public class ClienteMulticast {

    // OFFSET pessoal (secao 3.3 do roteiro) - deve ser o MESMO do servidor
    static final int OFFSET = 81;

    static final String GRUPO = "230.0.0.1";

    public static void main(String[] args) throws Exception {
        int porta = 4446 + OFFSET;
        InetAddress grupo = InetAddress.getByName(GRUPO);

        // Nome opcional so para identificar o cliente nos logs: java ClienteMulticast Aluno-1
        String nome = args.length > 0 ? args[0] : "Aluno";

        // ADAPTACAO PARA macOS
        // O roteiro usa NetworkInterface.getByInetAddress(InetAddress.getLocalHost()),
        // que no macOS costuma devolver null (o hostname nem sempre resolve para uma
        // interface real) e faz o joinGroup lancar excecao. Aqui a interface e
        // escolhida testando as que estao ativas e suportam multicast.
        NetworkInterface interfaceRede = escolherInterface();

        if (interfaceRede == null) {
            System.err.println("ERRO: nenhuma interface de rede com suporte a multicast foi encontrada.");
            System.err.println("Verifique se o Wi-Fi esta ligado e tente novamente.");
            listarInterfaces();
            return;
        }

        try (MulticastSocket socket = new MulticastSocket(porta)) {
            socket.setReuseAddress(true);

            SocketAddress enderecoGrupo = new InetSocketAddress(grupo, porta);
            socket.joinGroup(enderecoGrupo, interfaceRede);

            System.out.println("[MULTICAST] " + nome + " entrou no grupo " + GRUPO + ":" + porta);
            System.out.println("[MULTICAST] Interface usada: " + interfaceRede.getName()
                    + " (" + interfaceRede.getDisplayName() + ")");
            System.out.println("[MULTICAST] Aguardando avisos... (Ctrl+C para sair)");
            System.out.println();

            byte[] buffer = new byte[1024];
            while (true) {
                DatagramPacket pacote = new DatagramPacket(buffer, buffer.length);
                socket.receive(pacote);
                String mensagem = new String(
                        pacote.getData(), 0, pacote.getLength(), StandardCharsets.UTF_8);
                System.out.println("[" + nome + "] recebeu -> " + mensagem);
            }
        }
    }

    /**
     * Devolve a primeira interface ativa que suporta multicast e nao e loopback.
     * Em ultimo caso aceita o loopback, util quando a maquina esta sem rede.
     */
    private static NetworkInterface escolherInterface() throws SocketException {
        NetworkInterface loopback = null;

        for (NetworkInterface ni : Collections.list(NetworkInterface.getNetworkInterfaces())) {
            if (!ni.isUp() || !ni.supportsMulticast()) continue;
            if (ni.isLoopback()) {
                loopback = ni;
                continue;
            }
            // Precisa ter um endereco IPv4 configurado para valer como candidata
            for (InterfaceAddress addr : ni.getInterfaceAddresses()) {
                if (addr.getAddress() instanceof Inet4Address) {
                    return ni;
                }
            }
        }
        return loopback;
    }

    private static void listarInterfaces() throws SocketException {
        System.err.println("--- interfaces detectadas ---");
        for (NetworkInterface ni : Collections.list(NetworkInterface.getNetworkInterfaces())) {
            System.err.printf("  %-8s up=%-5s multicast=%-5s loopback=%s%n",
                    ni.getName(), ni.isUp(), ni.supportsMulticast(), ni.isLoopback());
        }
    }
}
